### PR TITLE
💪Revamp selector

### LINKED ISSUE #189: 💪Revamp selector
Our initial selector logic was an MVP, it is time to fix various problems:

A. 😴wait 30s for weak matches
B. 👴missing ancestor information
C. 😓hardcoded attributes (need to keep whitelisting keys #177 #172 #192)

### PR DESCRIPTION
**Pull request by jperl**
Resolves #189

To Do

- [x] Serialize node w/ ancestors -> dom string
- [x] Parse dom string -> doc w/ html-parse-string
- [x] Compare code
- [x] findElement
- [x] Change Recorder target to new serialization
- [x] Update formatIt etc to use new target
- [x] `findCss` & `find(selector)` & `find({ selector })` w/ tests
- [x] `findHtml` & `findElement({ html })` w/ tests
- [x] Replace html / body with simple elements, test `findHtml` matches & update fixtures
- [x] Fix tests
- [x] Serialize to html before saving workflow file
- [x] `findText` & `findElement({ text })` w/ tests

### CODE REVIEW

**Comment by jperl** [comment]
```
// make Locator compatible with Step
type Locator =  { html?, page?, selector?, text? };

find(selector | locator);
findHtml();
findText("my text");
findSelector();

click(locator || Promise<Element>);
click(findText("my text"));
click({ text: "hello" });
```

File: packages/browser/src/find/findHtml.ts [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+import { ElementHandle, Page } from "puppeteer";
+
+export type HtmlSelector = string | DocSelector;
+
+export const findHtml = async (
+  page: Page,
+  selector: HtmlSelector,
+  options: FindOptions
+): Promise<ElementHandle<Element> | null> => {
+  // TODO logging the html would be prettier...
```
Comment (flaurida) [inline packages/browser/src/find/findHtml.ts]: do this now or later?

File: packages/browser/src/find/findHtml.ts [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+      ? // convert the html to a document
+        { node: htmlToDoc(selector), ancestors: [] }
+      : selector;
+
+  const jsHandle = await page.evaluateHandle(
+    (docSelector, options) => {
+      const qawolf: QAWolfWeb = (window as any).qawolf;
+      return qawolf.find.findHtml(docSelector, options);
+    },
+    docSelector as any,
```
Comment (flaurida) [inline packages/browser/src/find/findHtml.ts]: do we have to do `as any`?

File: packages/browser/src/find/findCss.ts [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+
+export const findCss = async (
+  page: Page,
+  selector: string,
+  options: FindOptions
+): Promise<ElementHandle<Element> | null> => {
+  logger.verbose(`findCss: ${selector}`);
+
+  const jsHandle = await page.evaluateHandle(
+    (docSelector, options) => {
```
Comment (flaurida) [inline packages/browser/src/find/findCss.ts]: not a docSelector

File: packages/web/src/find/compare.ts [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+        childComparison.matches.map(key => `children[${index}].${key}`)
+      );
+      result.total += childComparison.total;
+    });
+  }
+
+  return result;
+};
+
+export const matchTarget = (
```
Comment (flaurida) [inline packages/web/src/find/compare.ts]: matchDocSelector would be more appropriate

File: packages/browser/src/find/findHtml.ts (hunk 2 above)
Comment (jperl) [reply packages/browser/src/find/findHtml.ts]: Yes because it relies on it being serializable

File: packages/browser/src/find/findHtml.ts (hunk 1 above)
Comment (jperl) [reply packages/browser/src/find/findHtml.ts]: Done

File: packages/web/src/find/compare.ts (hunk 4 above)
Comment (jperl) [reply packages/web/src/find/compare.ts]: Done

File: packages/browser/src/find/findCss.ts (hunk 3 above)
Comment (jperl) [reply packages/browser/src/find/findCss.ts]: Done

**Pull request merged**